src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs

```diff
@@ -386,5 +386,77 @@ TarEntryType.RegularFile or
+            // note: these are ordered child to parent.
+            while (directoryModificationTimes.TryPop(out (string Path, DateTimeOffset Modified) item))
+            {
+                AttemptDirectorySetLastWriteTime(item.Path, item.Modified);
+            }
+        }
+
+        public static void UpdatePendingModificationTimes(Stack<(string, DateTimeOffset)> directoryModificationTimes, string fullPath, DateTimeOffset modified)
+        {
+            // We can't set the modification time when we create the directory because extracting entries into it
+            // will cause that time to change. Instead, we track the times to set them later.
+
+            // We take into account that regular tar files are ordered:
+            // when we see a new directory which is not a child of the previous directory
+            // we can set the parent directory timestamps, and stop tracking them.
+            // This avoids having to track all directory entries until we've finished extracting the entire archive.
+            while (directoryModificationTimes.TryPeek(out (string Path, DateTimeOffset Modified) previous) &&
+                   !IsChildPath(previous.Path, fullPath))
+            {
+                directoryModificationTimes.TryPop(out previous);
+                AttemptDirectorySetLastWriteTime(previous.Path, previous.Modified);
+            }
+
+            directoryModificationTimes.Push((fullPath, modified));
+        }
+
+        private static bool IsChildPath(string parentFullPath, string childFullPath)
+        {
+            // Both paths may end with an additional separator.
+
+            // Verify that either the parent path ends with a separator
+            // or the child path has a separator where the parent path ends.
+            if (IsDirectorySeparatorChar(parentFullPath[^1]))
+            {
+                // The child needs to be at least a char longer than the parent for the name.
+                if (childFullPath.Length <= parentFullPath.Length)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                // The child needs to be at least 2 chars longer than the parent:
+                // one for the separator, and one for the name.
+                if ((childFullPath.Length < parentFullPath.Length + 2) ||
+                    !IsDirectorySeparatorChar(childFullPath[parentFullPath.Length]))
+                {
+                    return false;
+                }
+            }
+
+            return childFullPath.StartsWith(parentFullPath, PathInternal.StringComparison);
+
+            // We don't need to check for AltDirectorySeparatorChar, full paths are normalized to DirectorySeparatorChar.
+            static bool IsDirectorySeparatorChar(char c)
+                => c == Path.DirectorySeparatorChar;
+        }
+
+        private static void AttemptDirectorySetLastWriteTime(string fullPath, DateTimeOffset lastWriteTime)
```

Comment:
This is copied and changed for directories from:

https://github.com/dotnet/runtime/blob/e80ef8638b6d3f42add1bb046e032b2eff417473/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs#L552-L562